tensorzero-optimizers/src/gepa/pareto.rs

```diff
@@ -0,0 +1,2511 @@
+            };
+            return Err(Error::new(ErrorDetails::InternalError {
+                message: format!(
+                    "Validation scores contain {} datapoint(s) not in frontier's datapoint_ids: {unexpected_list}. \
+                     Layout must remain constant across iterations.",
+                    unexpected.len()
+                ),
+            }));
+        }
+
+        // Validate that all metric names are known in optimize_directions
+        let unknown_metrics: HashSet<String> = all_scores
+            .values()
+            .flat_map(|datapoint_scores| datapoint_scores.values())
+            .flat_map(|scores| scores.keys().cloned())
+            .filter(|metric| !self.optimize_directions.contains_key(metric))
+            .collect();
+
+        if !unknown_metrics.is_empty() {
+            let mut unknown_list: Vec<_> = unknown_metrics.iter().collect();
+            unknown_list.sort();
+            let formatted_list = if unknown_list.len() <= 5 {
+                unknown_list
+                    .iter()
+                    .map(|s| s.as_str())
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            } else {
+                format!(
+                    "{} and {} more",
+                    unknown_list[..5]
+                        .iter()
+                        .map(|s| s.as_str())
+                        .collect::<Vec<_>>()
+                        .join(", "),
+                    unknown_list.len() - 5
+                )
+            };
+            return Err(Error::new(ErrorDetails::InternalError {
+                message: format!(
+                    "Validation scores contain {} evaluator(s) not in frontier's evaluator set: {formatted_list}. \
+                     Evaluator layout must remain constant across iterations.",
+                    unknown_metrics.len()
+                ),
+            }));
+        }
+
+        // Create sorted metric directions for deterministic objective vector ordering
+        let mut sorted_directions: Vec<(&String, &MetricConfigOptimize)> =
+            self.optimize_directions.iter().collect();
+        sorted_directions.sort_by_key(|(name, _)| *name);
+
+        tracing::debug!(
+            "Step 1: Building instance-wise Pareto sets for {} datapoints (union across all variants)",
+            self.datapoint_ids.len()
+        );
+
+        // Step 1: Build instance-wise Pareto sets P*[i] for each datapoint
+        // Original paper: P*[i] = {variants with max score on instance i} (single evaluator)
+        // Our extension: P*[i] = {Pareto non-dominated variants on instance i} (multiple evaluators)
+        let mut instance_pareto_sets: HashMap<String, HashSet<String>> = HashMap::new();
+
+        for datapoint_id in &self.datapoint_ids {
+            // Collect scores for this datapoint across all variants
+            let mut instance_scores: Vec<(String, HashMap<String, Option<f32>>)> = Vec::new();
```

Comment:
use some of the new types to make it obvious what the data structure here is; it is not clear from inspection.